### PR TITLE
feat: add calltree creation method

### PR DESCRIPTION
**Pull request by NotPeopling2day**
### What I did

Adds a method for creating a calltree from a transaction trace. Implementing plugins, like hardhat, can request the trace from a transaction hash and pass the trace back to create the calltree.

fixes: https://github.com/ApeWorX/ape/issues/347

### Checklist

- [ ] Passes all linting checks (pre-commit and CI jobs)
- [ ] New test cases have been added and are passing
- [ ] Documentation has been updated
- [ ] PR title follows [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) standard (will be automatically included in the changelog)

### CODE REVIEW

File: evm_trace/base.py [hunk 1]
```diff
@@ -54,3 +56,121 @@ class CallTreeNode(BaseModel):
...
+    """
+    Creates a CallTreeNode from a given transaction trace.
+
+    Args:
+        trace (Iterator[TraceFrame]): Iterator of transaction trace frames.
+        show_internal (bool): Boolean whether to display internal calls. Defaulted to False.
+        root_node_kwargs (dict): Keyword argments passed to the root ``CallTreeNode``.
+
+    Returns:
+        CallTreeNode: Call tree of transaction trace.
```
Comment (unparalleled-js) [inline evm_trace/base.py:73]:
```suggestion
        :class:`~evm_trace.base.CallTreeNode`: Call tree of transaction trace.
```

File: evm_trace/base.py [hunk 2]
```diff
@@ -54,3 +56,121 @@ class CallTreeNode(BaseModel):
...
+    Args:
+        offset (HexBytes): Offset byte location in memory.
+        size (HexBytes): Number of bytes to return.
+        memory (List[HexBytes]): Memory stack.
+
+    Returns:
+        HexBytes: Byte value from memory stack.
+    """
+
+    size_int = int(size.hex(), 16)
```
Comment (unparalleled-js) [inline evm_trace/base.py:96]: You could optionally use `to_int` from `eth_utils` here

File: evm_trace/base.py [hunk 3]
```diff
@@ -54,3 +56,121 @@ class CallTreeNode(BaseModel):
...
+    start_word = math.floor(offset_int / 32)
+    # Compute the word that contains the last byte
+    stop_word = math.ceil((offset_int + size_int) / 32)
+
+    byte_slice = b""
+
+    for word in memory[start_word:stop_word]:
+        byte_slice += word
+
+    return HexBytes(byte_slice[(offset_int % 32) : size_int])  # noqa: E203
```
Comment (unparalleled-js) [inline evm_trace/base.py:113]: If you don't want to use `# noqa: E203`, you can set a variable first:

```python
offset_index = offset_int % 32
return HexBytes(byte_slice[offset_index: size_int])
```

File: evm_trace/base.py [hunk 4]
```diff
@@ -54,3 +56,121 @@ class CallTreeNode(BaseModel):
...
+) -> CallTreeNode:
+
+    if show_internal:
+        raise NotImplementedError()
+
+    node = CallTreeNode(**node_kwargs)
+
+    for frame in trace:
+        if frame.op in ("CALL", "DELEGATECALL", "STATICCALL"):
+            assert len(frame.stack) > 6
```
Comment (unparalleled-js) [inline evm_trace/base.py:127]: What is the nature of this `assert`? Keep in mind it is bad practice to use `assert` statements in production code. They are more suited for tests or local scripts. Use `exceptions` or write test cases to handle improper states.

File: evm_trace/base.py [hunk 5]
```diff
@@ -54,3 +56,121 @@ class CallTreeNode(BaseModel):
...
+
+        elif frame.op in ("RETURN", "REVERT"):
+            node.returndata = _extract_memory(
+                offset=frame.stack[-1], size=frame.stack[-2], memory=frame.memory
+            )
+            # TODO: Handle "execution halted" vs. gas limit reached
+            node.failed = frame.op == "REVERT"
+            break
+
+        # TODO: Handle invalid opcodes (`node.failed = True`)
```
Comment (unparalleled-js) [inline evm_trace/base.py:172]: There are quite a lot of `TODO` statements... Are we sure this PR is ready?

File: evm_trace/base.py (hunk 4 above)
Comment (NotPeopling2day) [reply evm_trace/base.py:127]: Thanks, I think that was left over from @fubuloubu and I debugging. I'll remove it!

File: evm_trace/base.py (hunk 5 above)
Comment (NotPeopling2day) [reply evm_trace/base.py:172]: It's not ready for production but it helps us to move forward with the critical parts. We'll then work on tweaking it before moving it out of alpha.

File: evm_trace/base.py (hunk 4 above)
Comment (NotPeopling2day) [reply evm_trace/base.py]: Updated with [4f8c476](https://github.com/ApeWorX/evm-trace/pull/4/commits/4f8c4768d925d448403736c8ce76490be1c9e3ca)

File: evm_trace/base.py (hunk 3 above)
Comment (NotPeopling2day) [reply evm_trace/base.py]: Updated with [4f8c476](https://github.com/ApeWorX/evm-trace/pull/4/commits/4f8c4768d925d448403736c8ce76490be1c9e3ca)

File: evm_trace/base.py (hunk 2 above)
Comment (NotPeopling2day) [reply evm_trace/base.py]: Updated with [4f8c476](https://github.com/ApeWorX/evm-trace/pull/4/commits/4f8c4768d925d448403736c8ce76490be1c9e3ca)

File: evm_trace/base.py (hunk 1 above)
Comment (NotPeopling2day) [reply evm_trace/base.py]: Updated with [4f8c476](https://github.com/ApeWorX/evm-trace/pull/4/commits/4f8c4768d925d448403736c8ce76490be1c9e3ca)

File: evm_trace/base.py [hunk 6]
```diff
@@ -54,3 +57,121 @@ class CallTreeNode(BaseModel):
...
+
+    return _create_node_from_call(
+        trace=trace,
+        show_internal=show_internal,
+        **root_node_kwargs,
+    )
+
+
+def _extract_memory(offset: HexBytes, size: HexBytes, memory: List[HexBytes]) -> HexBytes:
+    """
```
Comment (unparalleled-js) [inline evm_trace/base.py:85]: In general, you don't need "proper" documentation for methods prefixed with `_` (internal or private methods).
It is sometimes nice to know what it's doing, but always feel free to omit or be more brief

File: evm_trace/base.py [hunk 7]
```diff
@@ -54,3 +57,121 @@ class CallTreeNode(BaseModel):
...
+            child_node_kwargs["depth"] = frame.depth
+            # TODO: Validate gas values
+            child_node_kwargs["gas_limit"] = int(frame.stack[-1].hex(), 16)
+            child_node_kwargs["gas_cost"] = frame.gas_cost
+
+            if frame.op == "CALL":
+                child_node_kwargs["call_type"] = CallType.MUTABLE
+                child_node_kwargs["value"] = int(frame.stack[-3].hex(), 16)
+                child_node_kwargs["calldata"] = _extract_memory(
+                    offset=frame.stack[-4], size=frame.stack[-5], memory=frame.memory
```
Comment (unparalleled-js) [inline evm_trace/base.py:141]: I am going to take your word that this is all doing what it needs to be doing!
A reference to the official docs describing the domain knowledge here would be nice.

File: pyproject.toml [hunk 8]
```diff
@@ -35,7 +35,7 @@ markers = "fuzzing: Run Hypothesis fuzz test suite"
 line_length = 100
 force_grid_wrap = 0
 include_trailing_comma = true
-known_third_party = ["hexbytes", "pydantic", "pytest", "setuptools"]
+known_third_party = ["eth_utils", "hexbytes", "pydantic", "pytest", "setuptools"]
```
Comment (unparalleled-js) [inline pyproject.toml:38]: You can delete these and remove `seed-isort` from `pre-commit`.

It interferes with `isort` and slows it down. it is deprecated and should no longer be used.

See here: https://github.com/asottile-archive/seed-isort-config

**Review by unparalleled-js — APPROVED** [review]
LGTM

File: evm_trace/base.py (hunk 7 above)
Comment (fubuloubu) [reply evm_trace/base.py:141]: mostly from here: https://www.evm.codes/

Comment (NotPeopling2day) [reply evm_trace/base.py:145]: Added a link to the evm codes [f2cd2ef](https://github.com/ApeWorX/evm-trace/pull/4/commits/f2cd2efc98dbc46003082628584b66e7132b309d)

File: pyproject.toml (hunk 8 above)
Comment (NotPeopling2day) [reply pyproject.toml]: Removed isort [0c655d4](https://github.com/ApeWorX/evm-trace/pull/4/commits/0c655d459bf05a45bb2c97366acc8df7f940bd39)

**Review by fubuloubu — APPROVED** [review]

File: pyproject.toml (hunk 8 above)
Comment (NotPeopling2day) [reply pyproject.toml]: Removed isort [0c655d4](https://github.com/ApeWorX/evm-trace/pull/4/commits/0c655d459bf05a45bb2c97366acc8df7f940bd39)

**Pull request merged**